crux_core/src/capability/executor.rs

```diff
@@ -59,38 +90,88 @@ impl Spawner {
 
 // used in docs/internals/runtime.md
 // ANCHOR: arc_wake
-impl ArcWake for Task {
+impl ArcWake for NotifyTask {
     fn wake_by_ref(arc_self: &Arc<Self>) {
-        let cloned = arc_self.clone();
-        arc_self
-            .task_sender
-            .send(cloned)
-            .expect("unable to wake an async task, task sender channel is disconnected.")
+        let _ = arc_self.sender.send(arc_self.task_id);
+        // TODO should we report an error if send fails?
     }
 }
 // ANCHOR_END: arc_wake
 
+struct NotifyTask {
+    task_id: Uuid,
+    sender: Sender<Uuid>,
+}
+
 // used in docs/internals/runtime.md
 // ANCHOR: run_all
 impl QueuingExecutor {
     pub fn run_all(&self) {
-        // While there are tasks to be processed
-        while let Ok(task) = self.ready_queue.try_recv() {
-            // Unlock the future in the Task
-            let mut future_slot = task.future.lock().unwrap();
-
-            // Take it, replace with None, ...
-            if let Some(mut future) = future_slot.take() {
-                let waker = waker_ref(&task);
-                let context = &mut Context::from_waker(&waker);
-
-                // ...and poll it
-                if future.as_mut().poll(context).is_pending() {
-                    // If it's still pending, put it back
-                    *future_slot = Some(future)
-                }
+        // we read off both queues and execute the tasks we receive.
+        // Since either queue can generate work for the other queue,
+        // we read from them in a loop until we are sure both queues
+        // are exhaused
+        let mut did_some_work = true;
```

Comment:
There is a chance this might alter observable behaviour, since the running order of the tasks has potentially changed. Of course we shouldn't be relying on the order that effects arrive, but in practice at least in tests it is hard to avoid.

That could be avoided by having everything arrive on one queue (using an `enum`), not very pretty though.

Reply:
I think effectively random order of execution is a good thing - nothing should be relying on the ordering (beyond the causal one).